timemachine/integrator.py

```diff
@@ -88,6 +88,74 @@ def step(self, x, v):
         return new_x, new_v
 
 
+class VelocityVerletIntegrator(Integrator):
+    def __init__(self, force_fxn, masses, dt):
+        self.dt = dt
+        self.masses = masses[:, np.newaxis]  # TODO: cleaner way to handle (n_atoms,) vs. (n_atoms, 3) mismatch?
+        self.force_fxn = force_fxn
+
+    def step(self, x, v):
+        # note: 2 calls to force_fxn per timestep
+        v_mid = v + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+        x = x + self.dt * v_mid
+        v = v_mid + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+
+        return x, v
+
+    def multiple_steps(self, x, v, n_steps=1000):
+        # note: intermediate timesteps are staggered
+        #    xs[0], vs[0] = x_0, v_0
+        #    xs[1], vs[1] = x_2, v_{1.5}
+        #    xs[2], vs[2] = x_3, v_{2.5}
+        #    ...
+        #    xs[T-1], vs[T-1] = x_T, v_{T-0.5}
+        #    xs[T],   vs[T]   = x_T, v_T
+
+        # note: reorders loop slightly to avoid ~n_steps extraneous calls to force_fxn
+
+        zs = [(x, v)]
+
+        # initialize traj
+        v = v + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+        x = x + self.dt * v
+
+        # run n_steps-1 steps
+        for t in range(n_steps - 1):
+            v = v + (self.dt / self.masses) * self.force_fxn(x)
+            x = x + self.dt * v
+
+            zs.append((x, v))
+
+        # finalize traj
+        v = v + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+
+        zs.append((x, v))
+
+        xs = np.array([x for (x, _) in zs])
+        vs = np.array([v for (_, v) in zs])
+        return xs, vs
+
+    def _update_via_fori_loop(self, x, v, n_steps=1000):
```

Comment:
Do we need this? Seems to be only intended for testing and ends up duplicating some code

Reply:
Not just intended for testing -- also intended to be usable in situations where https://github.com/proteneer/timemachine/blob/2f06d9586693f97abba07452cb91b895265ec704/timemachine/integrator.py#L91 is usable